File: LeetCode/Easy/0110-balanced-binary-tree/0110-balanced-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsBalanced(TreeNode root) {
        if (root == null) return true;

        var stack = new Stack<TreeNode>();
        var heights = new Stack<int>();
        var current = root;
        var lastVisited = (TreeNode)null;

        while (current != null || stack.Count > 0)
        {
            if (current != null)
            {
                stack.Push(current);
                current = current.left;
            }
            else
            {
                var peek = stack.Peek();
                if (peek.right != null && peek.right != lastVisited)
                {
                    current = peek.right;
                }
                else
                {
                    var node = stack.Pop();
                    lastVisited = node;

                    var rHeight = node.right == null
                        ? 0
                        : heights.Pop();
                    var lHeight = node.left == null
                        ? 0
                        : heights.Pop();
                    if (Math.Abs(lHeight - rHeight) > 1 )
                    {
                        return false;
                    }
                    heights.Push(Math.Max(lHeight, rHeight) + 1);
                }
            }
        }
        
        return true;
    }
}

File: LeetCode/Easy/0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        stack: list[TreeNode] = []
        heights = []
        current = root
        last_visited = None


        while stack or current:
            while current:
                stack.append(current)
                current = current.left
            
            node = stack[-1]                
            if node.right and node.right != last_visited:
                current = node.right
            else:
                right_depth = 0 if not node.right else heights.pop()
                left_depth = 0 if not node.left else heights.pop()
                if abs(right_depth - left_depth) > 1:
                    return False
                depth = max(left_depth, right_depth) + 1
                heights.append(depth)

                last_visited = stack.pop()

        return True



        